bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionFinderService.java

```diff
@@ -0,0 +1,86 @@
+ */
+package org.openhab.core.config.discovery.addon;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.addon.AddonInfo;
+import org.openhab.core.addon.AddonInfoProvider;
+import org.openhab.core.config.discovery.addon.finders.AddonSuggestionFinder;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+
+/**
+ * This is a {@link AddonSuggestionFinderService} which discovers suggested Addons for the user to install.
+ *
+ * @author Andrew Fiddian-Green - Initial contribution
+ */
+@NonNullByDefault
+@Component(immediate = true, service = AddonSuggestionFinderService.class, name = AddonSuggestionFinderService.SERVICE_NAME)
+public class AddonSuggestionFinderService implements AutoCloseable {
+
+    public static final String SERVICE_NAME = "addon-suggestion-finder-service";
+
+    private final Set<AddonInfoProvider> addonInfoProviders = ConcurrentHashMap.newKeySet();
+    private final List<AddonSuggestionFinder> addonSuggestionFinders = Collections.synchronizedList(new ArrayList<>());
+
+    @Activate
+    public AddonSuggestionFinderService() {
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    public void addAddonInfoProvider(AddonInfoProvider addonInfoProvider) {
+        addonInfoProviders.add(addonInfoProvider);
+    }
+
+    public void removeAddonInfoProvider(AddonInfoProvider addonInfoProvider) {
+        addonInfoProviders.remove(addonInfoProvider);
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    public void addAddonSuggestionFinder(AddonSuggestionFinder addonSuggestionFinder) {
+        addonSuggestionFinders.add(addonSuggestionFinder);
+    }
+
+    public void removeAddonSuggestionFinder(AddonSuggestionFinder addonSuggestionFinder) {
+        addonSuggestionFinders.remove(addonSuggestionFinder);
+    }
+
+    @Deactivate
+    @Override
+    public void close() throws Exception {
+        addonSuggestionFinders.clear();
+        addonInfoProviders.clear();
+    }
+
+    public Set<AddonInfo> getSuggestedAddons(@Nullable Locale locale) {
+        List<AddonInfo> candidates = addonInfoProviders.stream().map(p -> p.getAddonInfos(locale))
+                .flatMap(Collection::stream).toList();
+        addonSuggestionFinders.forEach(f -> f.setAddonCandidates(candidates));
```

Comment:
There is something going on here. setAddonCandidates will add all candidates, creating duplicates if they are already in the candidates list. Every time the REST endpoint is called, the list will grow.
Do you need to separately keep track, or could you just integrate it in the getSuggestedAddons call.

Reply:
Did you check this? Every time getSuggestedAddons is called (and that is when the REST api suggestions endpoint is called), this method calls setAddonCandidates for each AddonSuggestionFinder. setAddonCandidates calls addonCandidates.addAll(candidates) in BaseAddonSuggestionFinder, adding the same candidates it already has in the list again.

Reply:
Let me think about this. I think the setAddonsCandidates method is there so I could inject data for testing, without needing to do a full osgi setup. But it should at least certainly clear the prior entries in the list.

Reply:
> I think the setAddonsCandidates method is there so I could inject data for testing, without needing to do a full osgi setup. But it should at least certainly clear the prior entries in the list.

Wouldn't it be better to try to mock that part then.